design/one-pager-resource-reclaim-policy.md

```diff
@@ -0,0 +1,93 @@
+5. A user deletes the `ResourceClaim`. Because the `Managed Resource` has an `OwnerReference` to the `ResourceClaim`, the deletion of the `ResourceClaim` triggers the deletion of the `Managed Resource`.
+6. If the `Managed Resource` `reclaimPolicy` is set to `Retain`, the `Managed Resource` will be deleted, but the `External Resource` will persist (i.e. the S3 bucket will still exist in your AWS account). If the `reclaimPolicy` is set to `Delete` both the `Managed Resource` and the `External Resource` will be deleted.
+
+## Goals
+
+Reclaim policies exist in Crossplane in order to provide the ability for an `External Resource` to persist outside of the lifecycle of a `ResourceClaim` to which they are bound. However, the current status of reclaim policies is flawed in that it allows for the possibility that an `External Resource` that was created by Crossplane persists after the deletion of the `Managed Resource` that represents it. This leads to an inaccurate state representation for a system that should be able to serve as the single control plane across cloud providers. Reclaim policies should more closely represent the functionality implemented in Kubernetes Persistent Volumes by tightly coupling the lifecyle of `Managed Resources` with their corresponding `External Resource`. Motivation for this change is provided by the following core concepts:
+
+* Crossplane should always provide an accurate representation of resource state (including that of external resources)
+* `External Resources` created by Crossplane should always have a `Managed Resource` representation unless the `Managed Resource` is manually removed by an administrator (i.e. no orphaned `External Resources`)
+* A tightly coupled `External Resource` / `Managed Resource` pair should have the option to persist after the deletion of any `Resource Claim` with which they are bound (i.e. a `reclaimPolicy` set to `Retain`)
+
+A move to change reclaim policies in Crossplane to match that of Kubernetes Persistent Volumes would result in the policies dictating the relationship between `ResourceClaims` and `Managed Resources` instead of the current status of `ResourceClaims` and `External Resources`. The only tradeoff in functionality with this shift is that unbound `Managed Resources` may continue to exist after deletion of a the `ResourceClaim` that they reference if their `reclaimPolicy` is set to `Retain`. This is opposed to the current system where the `Managed Resource` is always deleted upon `ResourceClaim` deletion, despite its `reclaimPolicy`. However, the persistence of the `Managed Resources` actually serves as a feature as they are a reminder that an unbound `External Resource` (which may be billing you) exists as well. To clean up the retained unbound `Managed Resource` and `External Resource`, an administrator will manually delete the `Managed Resource` from Crossplane, then take action on the `External Resource`.
```

Comment:
Do you have any thoughts on how we should address the use case of wanting to remove an External Resource from Crossplane management under this proposal? i.e. The functionality we get from the current implementation of the reclaim policy.

Perhaps removing External Resource from Crossplane management is better framed as part of the discussion about importing existing external resources into Crossplane management; i.e. some way to "attach" existing external resources to Crossplane and "detach" them without deleting them?

Reply:
Could you clarify what is meant by removing an External Resource from Crossplane management? If you mean that we want to be able to have an External Resource persist but no longer have a Managed Resource representation in Crossplane, that should be possible if we mirror PV [`Retain` policy](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retain):

> The Retain reclaim policy allows for manual reclamation of the resource. When the PersistentVolumeClaim is deleted, the PersistentVolume still exists and the volume is considered “released”. But it is not yet available for another claim because the previous claimant’s data remains on the volume. An administrator can manually reclaim the volume with the following steps.
> 1. Delete the PersistentVolume. The associated storage asset in external infrastructure (such as an AWS EBS, GCE PD, Azure Disk, or Cinder volume) still exists after the PV is deleted.
> 2. Manually clean up the data on the associated storage asset accordingly.
> 3. Manually delete the associated storage asset, or if you want to reuse the same storage asset, create a new PersistentVolume with the storage asset definition.

Reply:
> If you mean that we want to be able to have an External Resource persist but no longer have a Managed Resource representation in Crossplane

Yeah, that's what I mean here.

> that should be possible if we mirror PV Retain policy

Are you sure about that? My understanding (based on my reading of your linked documentation and some recent experiments) is that there's no way to delete a `PersistentVolume` resource from Kubernetes without also deleting the actual external volume it represents.

Reply:
I believe (1) above would suggest that the external volume persists after deletion of the `PersistentVolume`. Am I interpreting that incorrectly?

Reply:
Huh. Nope, you're correct. I somehow read that line as the complete opposite of what it actually says.

Reply:
So:

* When a claimed PV has `reclaimPolicy: Delete` deleting the PVC deletes the PV and the actual external volume.
* When a claimed PV has `reclaimPolicy: Retain` deleting the PVC retains both the PV and the actual external volume.
* When an unclaimed PV is deleted, the actual external volume remains regardless of the reclaim policy.

Is the above a correct understanding of how Kubernetes PV reclaimPolicy works?

Reply:
That is my understanding.

Reply:
Note that PV's also allow you to [change reclaim policy](https://kubernetes.io/docs/tasks/administer-cluster/change-pv-reclaim-policy/) after creation.